src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs

```diff
@@ -64,16 +54,31 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument)
 
             visitor.VisitBlock(syntaxTree.Root);
 
+            // In each lowering piece above, namespaces were tracked. We render them here to ensure every
+            // lowering action has a chance to add a source location to a namespace.
```

Comment:
I would also mention here that the **closest** wins. We want to use the `using` from the main source file if possible.